src/main/java/io/temporal/samples/encryptedpayloads/CryptDataConverter.java

```diff
@@ -0,0 +1,193 @@
+  private static final Charset UTF_8 = StandardCharsets.UTF_8;
+
+  private DataConverter converter;
+
+  public CryptDataConverter(DataConverter converter) {
+    this.converter = converter;
+  }
+
+  private SecretKey getKey() {
+    // Key can be fetched from KMS or other secure storage.
+    return new SecretKeySpec("test-key-test-key-test-key-test!".getBytes(UTF_8), "AES");
+  }
+
+  private static byte[] getNonce(int size) {
+    byte[] nonce = new byte[size];
+    new SecureRandom().nextBytes(nonce);
+    return nonce;
+  }
+
+  private byte[] encrypt(byte[] plainData) throws Exception {
+    SecretKey key = this.getKey();
+    byte[] nonce = getNonce(GCM_NONCE_LENGTH_BYTE);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    byte[] encryptedData = cipher.doFinal(plainData);
+    byte[] result = ByteBuffer.allocate(nonce.length + encryptedData.length).put(nonce).put(encryptedData).array();
+
+    return result;
+  }
+
+  private byte[] decrypt(byte[] encryptedDataWithNonce) throws Exception {
+    SecretKey key = this.getKey();
+    ByteBuffer buffer = ByteBuffer.wrap(encryptedDataWithNonce);
+
+    byte[] nonce = new byte[GCM_NONCE_LENGTH_BYTE];
+    buffer.get(nonce);
+    byte[] encryptedData = new byte[buffer.remaining()];
+    buffer.get(encryptedData);
+
+    Cipher cipher = Cipher.getInstance(CIPHER);
+    cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(GCM_TAG_LENGTH_BIT, nonce));
+
+    return cipher.doFinal(encryptedData);
+  }
+
+  @Override
+  public <T> Optional<Payload> toPayload(T value) throws DataConverterException {
+    Optional<Payload> optionalPayload = converter.toPayload(value);
+
+    if (!optionalPayload.isPresent()) {
+      return optionalPayload;
+    }
+
+    Payload innerPayload = optionalPayload.get();
+
+    ByteString encoding;
+    try {
+      encoding = innerPayload.getMetadataOrThrow(EncodingKeys.METADATA_ENCODING_KEY);
+    } catch (Throwable e) {
+      throw new DataConverterException(e);
+    }
+
+    byte[] encryptedData;
+    try {
+      encryptedData = encrypt(innerPayload.getData().toByteArray());
+    } catch (Throwable e) {
+      throw new DataConverterException(e);
+    }
+
+    Payload encryptedPayload = Payload.newBuilder(innerPayload)
+      .setData(ByteString.copyFrom(encryptedData))
+      .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_RAW)
+      .putMetadata(EncodingKeys.METADATA_WRAPPED_ENCODING_KEY, encoding)
+      .build();
+
+    return Optional.of(encryptedPayload);
+  }
+
+  @Override
+  public <T> T fromPayload(Payload payload, Class<T> valueClass, Type valueType) {
+    ByteString encoding;
+    try {
+      encoding = payload.getMetadataOrThrow(EncodingKeys.METADATA_WRAPPED_ENCODING_KEY);
+    } catch (Throwable e) {
+      return converter.fromPayload(payload, valueClass, valueType);
```

Comment:
I wouldn't use exceptions for flow control

Reply:
So I should use `payload.getMetadataOrDefault(...)`, or what would you suggest? I was expecting `getMetadata(key)` but that's not a thing.

Reply:
I'll refactor to use `getMetadataMap` just feels weird as I only need one field.

Reply:
Updated in https://github.com/temporalio/samples-java/pull/82/commits/27fbb88dbe1238ba9ffac52b85b2067a20598170

Reply:
Another option is to use `getMetadataOrDefault` which is a little cheaper than `getMetadata`.